freqtrade/optimize/backtesting.py

```diff
@@ -651,26 +689,60 @@ def backtest(self, processed: Dict,
                         open_trade_count += 1
                         # logger.debug(f"{pair} - Emulate creation of new trade: {trade}.")
                         open_trades[pair].append(trade)
-                        LocalTrade.add_bt_trade(trade)
 
                 for trade in list(open_trades[pair]):
-                    # also check the buying candle for sell conditions.
-                    trade_entry = self._get_sell_trade_entry(trade, row)
-                    # Sell occurred
-                    if trade_entry:
+                    # 2. Process buy orders.
+                    order = trade.select_order('buy', is_open=True)
+                    if order and self._get_order_filled(order.price, row):
+                        order.order_filled_date = row[DATE_IDX]
+                        trade.open_order_id = None
+                        order.filled = order.amount
+                        order.status = 'closed'
+                        order.ft_is_open = False
+                        LocalTrade.add_bt_trade(trade)
+
+                    # 3. Create sell orders (if any)
+                    if not trade.open_order_id:
+                        self._get_sell_trade_entry(trade, row)  # Place sell order if necessary
+
+                    # 4. Process sell orders.
+                    order = trade.select_order('sell', is_open=True)
+                    if order and self._get_order_filled(order.price, row):
+                        trade.open_order_id = None
+                        order.order_filled_date = trade.close_date = row[DATE_IDX]
+                        order.filled = order.amount
+                        order.status = 'closed'
+                        order.ft_is_open = False
+                        trade.close(order.price, show_msg=False)
+
                         # logger.debug(f"{pair} - Backtesting sell {trade}")
                         open_trade_count -= 1
                         open_trades[pair].remove(trade)
-
                         LocalTrade.close_bt_trade(trade)
-                        trades.append(trade_entry)
+                        trades.append(trade)
                         if enable_protections:
                             self.protections.stop_per_pair(pair, row[DATE_IDX])
-                            self.protections.global_stop(tmp)
+                            self.protections.global_stop(current_time)
+
+                    # 5. Cancel expired buy/sell orders.
+                    for order in [o for o in trade.orders if o.ft_is_open]:
+                        timeout = self.config['unfilledtimeout'].get(order.side, 0)
```

Comment:
We will want to call `check_buy_timeout` here too (and the selling, obviously) ... no?

Reply:
* i'll have a look at extracting that detached from this PR - as it'll simplify the code overall anyway.

Reply:
#6275